src/mysql_protocol.erl

```diff
@@ -1230,37 +1294,93 @@ get_null_terminated_binary(<<0, Rest/binary>>, Acc) ->
 
--spec hash_password(Password :: iodata(), Salt :: binary()) -> Hash :: binary().
-hash_password(Password, Salt) ->
+-spec hash_password(AuthMethod :: binary(), Password :: iodata(), Salt :: binary()) -> Hash :: binary().
+hash_password(AuthMethod, Password, Salt) when not is_binary(Password) ->
+    hash_password(AuthMethod, iolist_to_binary(Password), Salt);
+hash_password(?authmethod_none, Password, Salt) ->
+    hash_password(?authmethod_mysql_native_password, Password, Salt);
+hash_password(?authmethod_mysql_native_password, <<>>, _Salt) ->
+    <<>>;
+hash_password(?authmethod_mysql_native_password, Password, Salt) ->
     %% From the "MySQL Internals" manual:
     %% SHA1( password ) XOR SHA1( "20-bytes random data from server" <concat>
     %%                            SHA1( SHA1( password ) ) )
-    %% ----
-    %% Make sure the salt is exactly 20 bytes.
-    %%
-    %% The auth data is obviously nul-terminated. For the "native" auth
-    %% method, it should be a 20 byte salt, so let's trim it in this case.
-    PasswordBin = case erlang:is_binary(Password) of
-        true -> Password;
-        false -> erlang:iolist_to_binary(Password)
-    end,
-    case PasswordBin =:= <<>> of
-        true -> <<>>;
-        false -> hash_non_empty_password(Password, Salt)
-    end.
-
--spec hash_non_empty_password(Password :: iodata(), Salt :: binary()) ->
-    Hash :: binary().
-hash_non_empty_password(Password, Salt) ->
-    Salt1 = case Salt of
-        <<SaltNoNul:20/binary-unit:8, 0>> -> SaltNoNul;
-        _ when size(Salt) == 20           -> Salt
-    end,
-    %% Hash as described above.
+    Salt1 = trim_salt(Salt),
     <<Hash1Num:160>> = Hash1 = crypto:hash(sha, Password),
     Hash2 = crypto:hash(sha, Hash1),
     <<Hash3Num:160>> = crypto:hash(sha, <<Salt1/binary, Hash2/binary>>),
-    <<(Hash1Num bxor Hash3Num):160>>.
+    <<(Hash1Num bxor Hash3Num):160>>;
+hash_password(?authmethod_caching_sha2_password, <<>>, _Salt) ->
+    <<>>;
+hash_password(?authmethod_caching_sha2_password, Password, Salt) ->
+    %% From https://dev.mysql.com/doc/dev/mysql-server/latest/page_caching_sha2_authentication_exchanges.html
+    %% (transcribed):
+    %% SHA256( password ) XOR SHA256( SHA256( SHA256( password ) ) <concat>
+    %%                                        "20-bytes random data from server" )
+    Salt1 = trim_salt(Salt),
+    <<Hash1Num:256>> = Hash1 = crypto:hash(sha256, Password),
+    Hash2 = crypto:hash(sha256, Hash1),
+    <<Hash3Num:256>> = crypto:hash(sha256, <<Hash2/binary, Salt1/binary>>),
+    <<(Hash1Num bxor Hash3Num):256>>;
+hash_password(?authmethod_sha256_password, Password, Salt) ->
+    %% sha256_password authentication is superseded by
+    %% caching_sha2_password.
+    hash_password(?authmethod_caching_sha2_password, Password, Salt);
+hash_password(UnknownAuthMethod, _, _) ->
+    error({auth_method, UnknownAuthMethod}).
+
+encrypt_password(AuthMethod, Password, Salt, PubKey, ServerVersion)
```

Comment:
A comment/edoc about what (which auth methods) the function is used for would be good though.